### ISSUE TITLE
[FR] new "undistorted" masks form

### ISSUE THREAD

**Issue by AlicVB** [darktable-org/darktable#8714]
I continue to try to solve drawn masks issues discussed in #8634 ...
- drawn masks are now fully "what you see is what you get" whatever the distortion, before or after, even for cloning.
- the issue in retouch when there was distortion modules place before is now fixed too

Correct me if I'm wrong, but the last point that remain "to be solved" is the "true circle" "true straight gradient" part : if you have some distortion activated and use the "circle", you will get a "distorted" circle (in most cases an ellipse). You see now before adding them, how exactly it will be distorted... but you don't get a "true circle" !
This is even more problematic for gradient : if you have carefully lens correct + perspective correction + set the horizon ok, you want to be able to have a "true straight" line for your gradient (there's a way to set the curvature of the gradient, but that's not always enough)

@aurelienpierre has come with something in #8673 but that has lot of limitations and drawbacks, esp. because that drastically limit the use of distortions later in the workflow...

Here's my proposal to solve that : 
circles : 
- add a new "true" circle form (not necessary a new button, that can be some ctrl-click, or whatever)
- this new form will in fact be a "standard" path mask.
- the points of the path will match exactly the circle circumference in "screen" reference at the moment you draw it (as if you has done it by yourself)
- the number of point of the path mask should be enough so the gap between points is low (to avoid strange things if distorted after)
- as a "standard" path mask, user will be able to enlarge/reduce the circle, change the borders size, etc...
- as a "standard" path mask, the "content" of the circle will remain the same even if you distort your image later (if you mask a car wheel and then distort the image so wheels becomes ellipses, the masks will be distorted as an ellipse to still match the wheel)
- of course, if some distortion are applied after, the circle will not be a circle anymore, but at least, when you have created it, you get what you ask for ;)

gradient : 
- same as for circles proposal
- will need a little more work as path masks should be closed for instance
- that won't change the rendering algo, as it already work with a set of points, not just a line
- I think it's even more critical than circles (but harder to implement)

There's also ellipses, but honestly, I wonder if it's worse the pain, as it would be far more complex, esp. to handle the radius editions afterward...

technically, I think it should be doable, but I'd like to have your opinion on this before starting to implement something... esp. @aurelienpierre @TurboGit @elstoc as you have already reacted to the mentioned discussions

**Comment by elstoc** [comment]
It doesn't sound like a bad idea, though I'm not sure I've ever needed that much precision without going straight to drawing with a path. Most of the issues with gradient can be resolved by curving it (though the precision of that curve isn't quite what I'd like - a single step modifies it a little much for my taste). Most of the issues with circle can be solved by using a path or an ellipse.

It can sometimes be a bit fiddly but these days with mask feathering, I usually don't need that much precision.

An alternative might be a "convert to path" option, though you'd need some sort of mouse shortcut to do it.

**Comment by aurelienpierre** [comment]
The idea is good, the implementation will be full of bugs. There are too many layers of coordinates remapping and too many possible edge-cases going in the way, also remember not all the possible transforms are truly invertible in there. Not to mention the performance issue. I don't see a way to make the algebra robust enough (hell, we don't even manage high DPI scaling properly, my 100% zoom level in darkroom at 4K is still actual 200%… see #3008).

#8673 is mostly intended to cut down through the bullshit and have a safe fallback available, but also to enable (raster) drawn masks and layers in the (near) future (painted dodge and burn, that kind of thing…).

**Comment by TurboGit** [comment]
@AlicVB : In other words if I understand correctly you propose a kind of boolean in forms that says : don't distort back (previous module in pipe) but only forward new module in pipe. Right ? I suppose this should not be too hard to implement as we just need to call backtransform and not the forward transform routine for such forms. If this is not your idea please correct me. TIA.

**Comment by elstoc** [comment]
No the proposal here as I understand it (correct me if I'm wrong) is to add an option to implement "circles" as paths (with multiple nodes around the edge instead of just the central point and radius) so that the pipe distortions don't distort the circle as much. Similarly for the other "simple" shapes.

**Comment by AlicVB** [comment]
@TurboGit : no, @elstoc is right, I really think that we need the full distortion mechanism in any cases if we don't want to get very unpredictable results if we change distortions afterward, and I really don't want that we loose this possibility for "usual" cases.

My proposal here is "just" to solve the "issue" : "I have selected the circle tool, so I *want* a true circle, whatever the distortions" but still inside the current distortion system...

Now honestly, the "issue" to solve is not a real problem for me, so if nobody strongly wants it, I would be pleased to forget the idea :)



> The idea is good, the implementation will be full of bugs

like for every change done in code ;) (ok, maybe a little more for my changes :-D)
For the coordinates errors through distortions, as explain in another PR, I would be curious about quantifying them. In my test they are perfectly accurate, even with strong distortions ! (and in any cases, we have to keep in mind that drawn masks have a margin of error in their precision in any cases)

**Comment by TurboGit** [comment]
I see you want to record it but with distortion compensated for the circle to look like a circle on screen. But indeed in this case any distortions added after (and not after in the pipe as I did understood) will distort the circle again. My first thinking is then... What the point of doing all this work if the circle can be distorted again later? Is this worth the effort ? Especially since here we are speaking on a new kind of circle based on the path tool. I would say why not if he circle was just done as a path tool and all control point of the path tool visible. But adding flags to show a path as a circle with just a movable center control point will clutter the path code, so we really want to be sure all this is worth the effort (and maintenance cost to pay on the long term).
